---
title: "Markdown"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

#Script information

This script corrects the t-statistics and p-values generated by the last script for bias and inflation using the R package bacon, to prevent spurious correlations from influencing the results.

Link to the bacon package article: https://genomebiology.biomedcentral.com/articles/10.1186/s13059-016-1131-9

#Setup
```{r}
#Set working directory.
setwd("/home/tjonkman/researchdrive/tjonkman/epigenetic-clocks")

#set library path.
.libPaths("/home/tjonkman/researchdrive/tjonkman/epigenetic-clocks/Packages")

#Load all necessary libraries.
library(bacon)

options(stringsAsFactors = F)
```

#Run bacon on t-statistics obtained from the TWAS, in order to correct for any residual inflation and bias of the t-statistics.
#NB: bacon will be run separately for each CpG. This is because CATE was run using a separate model for each CpG, using that CpG to predict the expression of all 14,000 genes. In other words, each CpG represents a single related set of t-statistics, generated from the same linear model.
```{r}
#Load matrix of t-statistics of clocks and reference CpGs.
load("Output/Output_07b_CATE_clocks_effect_sizes.RData")
load("Output/Output_07c_CATE_clocks_t_statistics.RData")

#Set random seed (for reproducibility).
set.seed(1)

#Run BACON.
bc.clocks <- bacon(effectsizes = es.clocks, teststatistics = t.clocks)

#Extract the BACON-adjusted t-statistics and p-values for both clocks and refset.
#es.clocks <- es(bc.clocks)
#se.clocks <- se(bc.clocks)
t.clocks <- tstat(bc.clocks)
p.clocks <- pval(bc.clocks)
```

#Inspect output for clocks.
```{r}
#Inspect inflations and biases.
inflations <- inflation(bc.clocks)
summary(inflations)
hist(inflations, 100)

biases <- bias(bc.clocks)
summary(biases)
hist(biases, 100)

#
#Plot the distributions of the most and least inflated CpGs, and of the most and least biased CpGs.
#

#Least inflated (closest to 1)
test <- abs(1-inflations)
idx <- order(test, decreasing = F)
inflations[idx[1:10]]
fit(bc.clocks, n=50, index = idx[1], xlim = c(-6,6), main = "Least inflated")
posteriors(bc.clocks, index = idx[1])

#Most inflation
idx <- order(inflations, decreasing = T)
inflations[idx[1:10]]
fit(bc.clocks, n=50, index = idx[1], xlim = c(-6,6), main = "Most inflation")
posteriors(bc.clocks, index = idx[1])

#Most deflation
idx <- order(inflations, decreasing = F)
inflations[idx[1:10]]
fit(bc.clocks, n=50, index = idx[1], xlim = c(-6,6), main = "Most deflation")
posteriors(bc.clocks, index = idx[1])


#Least biased
test <- abs(0-biases)
idx <- order(test, decreasing = F)
biases[idx[1:10]]
fit(bc.clocks, n=50, index = idx[1], xlim = c(-6,6), main = "Least bias")
posteriors(bc.clocks, index = idx[1])

#Most negative bias
idx <- order(biases, decreasing = F)
biases[idx[1:10]]
fit(bc.clocks, n=50, index = idx[1], xlim = c(-6,6), main = "Most negative bias")
posteriors(bc.clocks, index = idx[1])

#Most positive bias
idx <- order(biases, decreasing = T)
biases[idx[1:10]]
fit(bc.clocks, n=50, index = idx[1], xlim = c(-6,6), main = "Most positive bias")
posteriors(bc.clocks, index = idx[1])
```

#Save BACON output.
```{r}
save(bc.clocks, file = "Output/Output_08a.0_bacon_full_clocks.RData")
save(t.clocks, file = "Output/Output_08a.1_bacon_t_clocks.RData")
save(p.clocks, file = "Output/Output_08a.2_bacon_p_clocks.RData")
```

#Plot some correlations.
```{r}
#Set working directory.
setwd("/home/tjonkman/researchdrive/tjonkman/epigenetic-clocks")

#set library path.
.libPaths("/home/tjonkman/researchdrive/tjonkman/epigenetic-clocks/Packages")

library(SummarizedExperiment)
library(ggplot2)
load("Output/Output_03a_twas_data.RData")
load("Output/Output_08a.1_bacon_t_clocks.RData")
load("Output/Output_01a_ensembl_genes.RData")

t.min <- sort(t.clocks, decreasing = F)
t.max <- sort(t.clocks, decreasing = T)
head(t.min, 100)
head(t.max, 100)

load("Output/Output_09a_significant_associations_in_trans.RData")
sig.associations[1:5,1:5]

mvalues <- assay(mvalues.clocks)
mvalues[1:3,1:3]

#Calculate correlation matrix for 365 clock CpGs and 216 trans-genes.
subset.counts <- RIN.counts[rownames(sig.associations),]
subset.mvalues <- mvalues[colnames(sig.associations),]
dim(subset.counts)
dim(subset.mvalues)

cor.mat <- cor(t(subset.counts), t(subset.mvalues))
dim(cor.mat)
cor.mat[1:3,1:3]
```

```{r}
Plot.Cor <- function(gene, dec.idx, filename){
  
  expression <- (RIN.counts)[gene,]
  cpg <- names(which(cor.mat[gene,] == sort(cor.mat[gene,], decreasing = dec.idx)[1]))
  methylation <- mvalues[cpg,]
  
  plot.data <- data.frame(expression = expression, 
                                    methylation = methylation)
  head(plot.data)
  
  #Add correlation data for annotation on top of the plot.
  test <- cor.test(plot.data$expression, plot.data$methylation)
  test$estimate
  test$p.value
  
  annots <- data.frame(
    corr = paste0("r == ", round(test$estimate, digits = 2)),
    pval = test$p.value
    )
  
  #R shows extremely small p-values as 0. If this is the case, simply write "p < 10^-15".
  if(annots$pval < 1e-15){
    annots$pval <- "p < 10^-15"
  } else {
    annots$pval <- paste0("p == ", annots$pval)
  }
  annots
  
  ggplot(data = plot.data, mapping = aes(x = expression, y = methylation)) + 
    geom_point() +
    geom_smooth(method='lm', formula= y~x) +
    annotate("text", x = -3.5, y = 1+(max(plot.data$methylation)), label=annots$corr, parse=TRUE, size = 5, hjust = 0) +
    annotate("text", x = -3.5, y = 0.5+(max(plot.data$methylation)), label=annots$pval, parse=TRUE, size = 5, hjust = 0) +
    labs(x = paste0("Normalized ", GeneEns[gene]$symbol, " expression"), 
         y = paste0(cpg, " methylation (M-value)"))
  
  ggsave(height = 4, width = 6, file = filename)
}

#positive correlations.
Plot.Cor(gene = "ENSG00000114861", dec.idx = T, 
         filename = "test/example of positive correlation 1.png")

Plot.Cor(gene = "ENSG00000138795", dec.idx = T, 
         filename = "test/example of positive correlation 2.png")

Plot.Cor(gene = "ENSG00000167664", dec.idx = T, 
         filename = "test/example of positive correlation 3.png")

Plot.Cor(gene = "ENSG00000089692", dec.idx = T, 
         filename = "test/example of positive correlation 4.png")

Plot.Cor(gene = "ENSG00000163508", dec.idx = T, 
         filename = "test/example of positive correlation 5.png")

Plot.Cor(gene = "ENSG00000100450", dec.idx = T, 
         filename = "test/example of positive correlation 6.png")



#negative correlations.
Plot.Cor(gene = "ENSG00000114861", dec.idx = F, 
         filename = "test/example of negative correlation 1.png")

Plot.Cor(gene = "ENSG00000138795", dec.idx = F, 
         filename = "test/example of negative correlation 2.png")

Plot.Cor(gene = "ENSG00000167664", dec.idx = F, 
         filename = "test/example of negative correlation 3.png")

Plot.Cor(gene = "ENSG00000089692", dec.idx = F, 
         filename = "test/example of negative correlation 4.png")

Plot.Cor(gene = "ENSG00000163508", dec.idx = F, 
         filename = "test/example of negative correlation 5.png")

Plot.Cor(gene = "ENSG00000100450", dec.idx = F, 
         filename = "test/example of negative correlation 6.png")


#Find the strongest correlations, both positive and negative.
gene.maxes <- apply(cor.mat, 1, max)
gene.idx <- names(which(gene.maxes == max(gene.maxes)))
Plot.Cor(gene = gene.idx, dec.idx = T, 
         filename = "test/strongest positive correlation.png")

gene.mins <- apply(cor.mat, 1, min)
gene.idx <- names(which(gene.mins == min(gene.mins)))
Plot.Cor(gene = gene.idx, dec.idx = F, 
         filename = "test/strongest negative correlation.png")
```

```{r}
i <- 2
idx <- which(t.clocks == t.min[i])

gene.idx <- idx %% nrow(t.clocks)
cpg.idx <- which(t.clocks[gene.idx,] == t.min[i])

expression <- (RIN.counts)[gene.idx,]
methylation <- as.numeric(assay(mvalues.clocks[colnames(t.clocks)[cpg.idx],]))
plot.data <- data.frame(expression = expression, 
                                  methylation = methylation)
head(plot.data)

#Add correlation data for annotation on top of the plot.
test <- cor.test(plot.data$expression, plot.data$methylation)
test$estimate
test$p.value

annots <- data.frame(
  corr = paste0("r == ", round(test$estimate, digits = 2)),
  pval = test$p.value
  )

#R shows extremely small p-values as 0. If this is the case, simply write "p < 10^-15".
if(annots$pval < 1e-15){
  annots$pval <- "p < 10^-15"
} else {
  annots$pval <- paste0("p == ", annots$pval)
}
annots

ggplot(data = plot.data, mapping = aes(x = expression, y = methylation)) + 
  geom_point() +
  geom_smooth(method='lm', formula= y~x) +
  annotate("text", x = 2, y = 1*(max(plot.data$methylation)), label=annots$corr, parse=TRUE, size = 6, hjust = 0) +
  annotate("text", x = 2, y = 0.8*(max(plot.data$methylation)), label=annots$pval, parse=TRUE, size = 6, hjust = 0) +
  labs(x = paste0("Normalized ", GeneEns[rownames(RIN.counts)[gene.idx]]$symbol, " expression"), 
       y = paste0(names(cpg.idx), " methylation (M-value)"))

ggsave(height = 4, width = 6, file = "test/example of negative correlation.png")
```

#Session info.
```{r}
sessionInfo()
```